Add documentation for annotations resolving (#10592)
pydantic/pydantic: docs/internals/annotations_resolving.md

!!! note
    This section is part of the *internals* documentation, and is partly targeted to contributors.

> Pydantic heavily relies on [type hints][type hint] at runtime to build schemas for validation, serialization, etc.

Author: Python has a glossary entry identified by `type hint`. But unfortunately the mkdocs ecosystem has limited support for cross refs, ideally we should be explicit about the target. With Sphinx, it looks like:

```
:external+python:ref:`type hints <type hint>`
```

I think I'm going to propose a feature request to support more use cases as I'd like to use cross refs more and more

While type hints were primarily introduced for static type checkers (such as [Mypy] or [Pyright]), they are
accessible (and sometimes evaluated) at runtime. This means that the following would fail at runtime,
because `Node` has yet to be defined in the current module:

```python test="skip" lint="skip"
class Node:
    """Binary tree node."""

    # NameError: name 'Node' is not defined:
    def __init__(self, l: Node, r: Node) -> None:
        self.left = l
        self.right = r
```

To circumvent this issue, forward references can be used (by wrapping the annotation into quotes).

In Python 3.7, [PEP 563] introduced the concept of _postponed evaluation of annotations_, meaning
with the `from __future__ import annotations` [future statement], type hints are stringified by default:

```python requires="3.12" lint="skip"
from __future__ import annotations

from pydantic import BaseModel

class Foo(BaseModel):
    f: MyType
    # Without the future import, equivalent to:

>     # f: 'MyType'

Reviewer: Maybe this would be more clear as like "given the future import above, this is equivalent to f: 'MyType'" or something like that?

type MyType = int

print(Foo.__annotations__)
#> {'f': 'MyType'}
```

## The challenges of runtime evaluation

Static type checkers make use of the <abbr title="Abstract Syntax Tree">AST</abbr> to analyze the defined annotations.
Regarding the previous example, this has the benefit of being able to understand what `MyType` refers to when analyzing

> the class definition of `Foo`, even if it isn't yet defined at runtime.

Reviewer: ```suggestion
the class definition of `Foo`, even if `MyType` isn't yet defined at runtime.
```

However, for runtime tools such as Pydantic, it is more challenging to correctly resolve these forward annotations.
The Python standard library provides some tools to do so ([`typing.get_type_hints()`][typing.get_type_hints],
[`inspect.get_annotations()`][inspect.get_annotations]), but they come with some limitations. Thus, they are
being re-implemented in Pydantic with improved support for edge cases.

For a long time, Pydantic has tried to support many cases, resulting in a messy annotations evaluation logic

> (which would allow invalid use cases from a static type checking perspective to work). In 2.10, the internal logic

Reviewer: So, I fully agree with what you're saying here.

That being said, I don't think we should be quite this negative in the docs. Maybe something like:

As Pydantic as grown, it's adapted to support many edge cases requiring irregular patterns for annotation evaluation. Some of these use cases aren't necessarily sound from a static type checking perspective. In v2.10, the internal namespace management logic was refactored in an attempt to simplify and standardize annotation evaluation. Admittedly, backwards compatibility posed some challenges, and there is still some noticeable scar tissue in the codebase because of this. There's hope that...

was refactored to support more cases while still trying to be backwards compatible. There's a hope that [PEP 649]
(introduced in Python 3.14) will greatly simplify the process, especially when it comes to dealing with locals
of a function.

To evaluate forward references, Pydantic roughly follows the same logic as described in the documentation of the
[`typing.get_type_hints()`][typing.get_type_hints] function. That is, the built-in [`eval()`][eval] function is used

> by passing the forward reference, a global and a local namespace. The namespace fetching logic is defined in the

Reviewer: ```suggestion
by passing the forward reference, a global, and a local namespace. The namespace fetching logic is defined in the
```

sections below.

## Resolving annotations at class definition

The following example will be used as a reference throughout this section:

```python test="skip" lint="skip"
# module1.py:
type MyType = int

class Base:
    f1: 'MyType'

# module2.py:
from pydantic import BaseModel

from module1 import Base

type MyType = str

def inner() -> None:
    type InnerType = bool

    class Model(BaseModel, Base):
        type LocalType = bytes

        f2: 'MyType'
        f3: 'InnerType'
        f4: 'LocalType'
        f5: 'UnknownType'

    type InnerType2 = complex
```

> When the `Model` class is being built, different [namespaces][namespace] are in play. For each base class

Reviewer: ```suggestion
When the `Model` class is being built, different [namespaces][namespace] are at play. For each base class
```

         - The parent namespace of the class, if different from the globals described above. This is the
           [locals][frame.f_locals] of the frame where the class is being defined. For `Base`, because the class is being
           defined in the module directly, this namespace won't be used as it will result in the globals being used again.
           For `Model`, the parent namespace is the locals of the frame of `inner()`.
3. If the annotation failed to evaluate, it is kept as is, so that the model can be rebuilt at a later stage. This will
   be the case for `f5`.

The following table lists the resolved type annotations for every field, once the `Model` class has been created:

| Field name | Resolved annotation |
|------------|---------------------|
| `f1`       | [`int`][]           |
| `f2`       | [`str`][]           |
| `f3`       | [`bool`][]          |
| `f4`       | [`bytes`][]         |
| `f5`       | `'UnkownType'`      |

### Limitations and backwards compatibility concerns

While the namespace fetching logic is trying to be as accurate as possible, we still face some limitations:

> <div class="annotate" markdown>

Reviewer: Nice, didn't know you could do this! Preview looks great.

- The locals of the current class (`cls.__dict__`) may include irrelevant entries, most of them being dunder attributes.
  This means that the following annotation: `f: '__doc__'` will successfully (and unexpectedly) be resolved.
- When the `Model` class is being created inside a function, we keep a copy of the [locals][frame.f_locals] of the frame.
  This copy only includes the symbols defined in the locals when `Model` is being defined, meaning `InnerType2` won't be included
  (and will **not be** if doing a model rebuild at a later point!).
- To avoid memory leaks, we use [weak references][weakref] to the locals of the function, meaning some forward references might
  not resolve outside the function (1).

> - Locals of the function are only taken into account for Pydantic models, it does not apply to dataclasses, typed dictionaries
>   or named tuples.

Reviewer: ```suggestion
- Locals of the function are only taken into account for Pydantic models, but this pattern does not apply to dataclasses, typed dictionaries
  or named tuples.
```

>   or named tuples.

Reviewer: Can we make these sub-bullets of the above bullet given that they all refer to the function context?

</div>

1.  Here is an example:

    ```python test="skip" lint="skip"
    def func():
        A = int

        class Model(BaseModel):
            f: 'A | Forward'

        return Model

    Model = func()

    Model.model_rebuild(_types_namespace={'Forward': str})
    # pydantic.errors.PydanticUndefinedAnnotation: name 'A' is not defined
    ```

[](){#backwards-compatibility-logic}

For backwards compatibility reasons, and to be able to support valid use cases without having to rebuild models,
the namespace logic described above is a bit different when it comes to core schema generation. Taking the
following example:

```python
from dataclasses import dataclass

from pydantic import BaseModel

@dataclass
class Foo:
    a: 'Bar | None' = None

class Bar(BaseModel):
    b: Foo
```

Once the fields for `Bar` have been collected (meaning annotations resolved), the `GenerateSchema` class converts

> every field into a core schema. When it encounters another class defining fields (such as dataclasses), it will

Reviewer: ```suggestion
every field into a core schema. When it encounters another class like field type (such as a dataclass), it will
```
